x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java

```diff
@@ -0,0 +1,482 @@
+                                "] but previous value was [" + previousVal +"]");
+                        }
+                        values.add(val);
+                        previousVal = val;
+                        token = context.parser().nextToken();
+                    }
+                } else if (fieldName.equals(COUNTS_FIELD.getPreferredName())) {
+                    token = context.parser().nextToken();
+                    // should be an array
+                    ensureExpectedToken(XContentParser.Token.START_ARRAY, token, context.parser()::getTokenLocation);
+                    counts = new IntArrayList();
+                    token = context.parser().nextToken();
+                    while (token != XContentParser.Token.END_ARRAY) {
+                        // should be a number
+                        ensureExpectedToken(XContentParser.Token.VALUE_NUMBER, token, context.parser()::getTokenLocation);
+                        counts.add(context.parser().intValue());
+                        token = context.parser().nextToken();
+                    }
+                } else {
+                    throw new MapperParsingException("error parsing field [" +
+                        name() + "], with unknown parameter [" + fieldName + "]");
+                }
+                token = context.parser().nextToken();
+                level = maybeAddOrRemoveLevel(token, level);
+            }
+            level = 0;
+            if (values == null) {
+                throw new MapperParsingException("error parsing field ["
+                    + name() + "], expected field called [" + VALUES_FIELD.getPreferredName() + "]");
+            }
+            if (counts == null) {
+                throw new MapperParsingException("error parsing field ["
+                    + name() + "], expected field called [" + COUNTS_FIELD.getPreferredName() + "]");
+            }
+            if (values.size() != counts.size()) {
+                throw new MapperParsingException("error parsing field ["
+                    + name() + "], expected same length from [" + VALUES_FIELD.getPreferredName() +"] and " +
+                    "[" + COUNTS_FIELD.getPreferredName() +"] but got [" + values.size() + " != " + counts.size() +"]");
+            }
+            if (fieldType().hasDocValues()) {
+                BytesStreamOutput streamOutput = new BytesStreamOutput();
+                for (int i = 0; i < values.size(); i++) {
+                    int count = counts.get(i);
+                    if (count < 0) {
+                        throw new MapperParsingException("error parsing field ["
+                            + name() + "], ["+ COUNTS_FIELD + "] elements must be >= 0 but got " + counts.get(i));
+                    } else if (count > 0) {
+                        // we do not add elements with count == 0
+                        streamOutput.writeDouble(values.get(i));
+                        streamOutput.writeVInt(count);
```

Comment:
I'd suggest putting the count before the values, it might make it easier to better compress in the future by stealing bits of the count.